plugins/com.google.cloud.tools.eclipse.appengine.deploy.ui/src/com/google/cloud/tools/eclipse/appengine/deploy/ui/flexible/FlexDeployCommandHandler.java

```diff
@@ -39,14 +46,22 @@ protected DeployPreferencesDialog newDeployPreferencesDialog(Shell shell, IProje
   }
 
   @Override
-  protected StagingDelegate getStagingDelegate(IProject project) {
+  protected StagingDelegate getStagingDelegate(IProject project) throws CoreException {
     String appYamlPath = new FlexDeployPreferences(project).getAppYamlPath();
     IFile appYaml = project.getFile(appYamlPath);
     if (!appYaml.exists()) {
       throw new IllegalStateException("Invalid app.yaml path from project preferences.");
     }
     IPath appEngineDirectory = appYaml.getParent().getLocation();
-    return new FlexStagingDelegate(appEngineDirectory);
+
+    IFacetedProject facetedProject = ProjectFacetsManager.create(project);
+    if (AppEngineFlexWarFacet.hasFacet(facetedProject)) {
+      return new FlexStagingDelegate(appEngineDirectory);
+    } else if (AppEngineFlexJarFacet.hasFacet(facetedProject)
+        && MavenUtils.hasMavenNature(project)) {
+      return new FlexJarMavenProjectStagingDelegate(appEngineDirectory);
+    }
+    throw new IllegalStateException("BUG: command enabled for non-flex or non-Maven-flex projects");
```

Comment:
This should be two different exception messages then, one for no flex facet at all and the other for a flex jar facet without a maven nature.